### PR TITLE
Cleanups in libp2p-core in stable-futures branch

### PR DESCRIPTION
**Pull request by tomaka**
Removes a few left-overs, and lots of `Unpin` trait bounds replaced with `pin-project`.

### CODE REVIEW

**Comment by tomaka** [comment]
Note that, unless we do some design refactoring, the substream upgrade infrastructure has to require `Unpin` for the substream type, as we have to read/write to them, then give ownership to them to the upgrade function.

File: core/src/either.rs [hunk 1]
```diff
@@ -56,99 +57,75 @@ where
...
-{
-    fn read(&mut self, buf: &mut [u8]) -> Result<usize, IoError> {
-        match self {
-            EitherOutput::First(a) => a.read(buf),
-            EitherOutput::Second(b) => b.read(buf),
+    #[project]
+    fn poll_read(self: Pin<&mut Self>, cx: &mut Context, buf: &mut [u8]) -> Poll<Result<usize, IoError>> {
+        #[project]
+        match self.project() {
+            EitherOutput::First(a) => AsyncRead::poll_read(a, cx, buf),
```
Comment (twittner) [inline core/src/either.rs]: Here and elsewhere you switched to UFCS style. What is the motivation for this change and what is the advantage here over writing `a.poll_read(cx, buf)`?

Comment (tomaka) [reply core/src/either.rs]: It's just a habit from early days of `futures-preview` where it wasn't possible to write `a.poll_read`.
